### ISSUE TITLE
Feature: Splitting into Sub-batches

### ISSUE THREAD

**Issue by srush** [pytorch/text#252]
This one may be infeasible, but I would love to be able to subdivide a batch into subbatches (or even subranges over the time dimension). For instance, this would allow us to directly send a batch object to DataParallel. 

We have some code I would love to get rid of that loops through a batch and splits everything by batch or by time .

**Comment by jekbradbury** [comment]
This is a good idea. Honestly I somewhat regret the original decision to make `Batch` an object rather than just a dictionary, because the PyTorch folks have implemented support for things like DataParallel in such a way that they apply to native Python container types but not custom ones. Perhaps there's a way to duck-type and pretend to be sufficiently dictionary-like?

**Comment by srush** [comment]
I mostly agree. Particularly since it is natural to use fields as a dictionary. 

The one nice thing about objects is that we could maybe have subfeatures. But that may be overkill.

**Comment by jekbradbury** [comment]
Yes, that (and attaching metadata) was the original motivation. But nested dicts would have been just fine, and metadata could just have been underscore-prefixed keys or something.
In some ways this gives us nice abstraction: we can get the `len` of a batch (it's the batch size), or `repr` it for pretty-printing. But what we'd really like is to be able to pass the batch to code that doesn't know about the type, and the set of built-in functions Python lets us overload is way too limited to accomplish that.

**Comment by jekbradbury** [comment]
I've been playing too much with [Julia](https://github.com/julialang/julia) recently, so now I feel like nothing else (least of all Python) has good enough generics...

**Comment by srush** [comment]
@malmaud is our group's resident "everyone should switch to julia" troll

**Comment by malmaud** [comment]
Although now I just use Python and pytorch...

**Comment by apaszke** [comment]
@jekbradbury you could inherit from `dict` or implement `__instancecheck__`, right?

**Comment by jekbradbury** [comment]
I'm not familiar with `__instancecheck__` but it looks straightforward enough. I'll try inheriting from dict  first and see if it can provide the best of both worlds here, including a potentially nonbreaking upgrade path.
